openpype/modules/kitsu/plugins/publish/collect_kitsu_entities.py

```diff
@@ -13,61 +17,69 @@ class CollectKitsuEntities(pyblish.api.ContextPlugin):
 
     def process(self, context):
 
-        asset_data = context.data["assetEntity"]["data"]
-        zou_asset_data = asset_data.get("zou")
+        # Get all needed names
```

Comment:
I'm afraid this is not right. Each instance may have different asset/task.
Few points during publishing:
- `context.data` always have `"projectName"`
- `context.data` may have `"asset"` and `"task"` if is in DCC where is "current context" (workfile opened on specific task)
- `instance.data` can have the same `"asset"` and `"task"` as `context.data` but not necessarily
    - each instance should have it's own information about kitsu entity because each may have different one

BTW at this point each instance already have `instance.data["assetEntity"]` so you don't have to query them here.

Reply:
I was following directions by Jakub Trllo on how to access the asset and task name from Tray Publisher.
From Fusion `context.data.get("asset")` exists but not from Tray Publisher, thus the check for `instance.data.get("asset")` later if asset_name doesn't exist.

Is fetching the asset and task name from `instance.data["assetEntity"]` more correct in this instance or how should this part of the code be written more correctly?

Reply:
Would love to get more clarification on this so I can fix what's wrong @iLLiCiTiT

Reply:
I've tried to explain this on Discord https://discord.com/channels/517362899170230292/517382145552154634/1073363775903780915

For GitHub record: Any asset/task information should be taken from `instance.data` except some cases when data are context based (e.g. dynamic creation of instances). Result of this plugin should be that each `instance.data` have filled kitsu entities based on their values of `"asset"` and `"task"`. And entities from `context.data` are not much relevant.

```python
kitsu_project = None

kitsu_entities_by_id = {}
for instance in context:
    asset_doc = instance.data.get("assetEntity")
    task_name = instance.data.get("task")
    if not asset_doc:
        continue

    zou_asset_data = asset_doc["data"].get("zou")
    if not zou_asset_data:
        raise AssertionError("Zou asset data not found in OpenPype!")

    if kitsu_project is None:
        kitsu_project = gazu.project.get_project(zou_asset_data["project_id"])
        if not kitsu_project:
            raise AssertionError("Project not found in kitsu!")

    entity_type = zou_asset_data["type"]
    kitsu_id = zou_asset_data["id"]
    kitsu_entity = kitsu_entities_by_id.get(kitsu_id)
    if not kitsu_entity:
        if entity_type == "Shot":
            kitsu_entity = gazu.shot.get_shot(kitsu_id )
        else:
            kitsu_entity = gazu.asset.get_asset(kitsu_id )
        kitsu_entities_by_id[kitsu_id] = kitsu_entity

    if not kitsu_entity:
        raise AssertionError(f"{entity_type} not found in kitsu!")
    instance.data["kitsu_entity"] = kitsu_entity

    if not task_name:
        continue
    zou_task_data = asset["data"]["tasks"][task_name].get("zou")
    self.log.debug(f"Collected zou task data: {zou_task_data}")
    if not zou_task_data:
        kitsu_task_type = gazu.task.get_task_type_by_name(task_name)
        if not kitsu_task_type:
            raise AssertionError(
                "Task type {} not found in Kitsu!".format(task_name)
            )
        continue
    kitsu_task_id = zou_task_data["id"]
    kitsu_task = kitsu_entities_by_id.get(kitsu_task_id)
    if not kitsu_task:
        kitsu_task = gazu.task.get_task(zou_task_data["id"])
        kitsu_entities_by_id[kitsu_task_id] = kitsu_task

    if not kitsu_task:
        raise AssertionError("Task not found in kitsu!")
    instance.data["kitsu_task"] = kitsu_task
    self.log.debug(f"Collect kitsu task: {kitsu_task}")
       

context.data["kitsu_project"] = kitsu_project
```

Reply:
Thank you so much for this! It really helped me a lot to understand the communication within OP.
Fixing all codes within this PR to work as this